### PR TITLE
[jaeger] Add CronJob to delete ElasticSearch indices older than n days

### PR DESCRIPTION
**Pull request by mehta-ankit**
Signed-off-by: Ankit Mehta <ankit.mehta@appian.com>

* Adds a CronJob resource as part of helm chart install for Jaeger Chart. This runs every day at 23:55 and deletes Elasticsearch indices older than `n` days (where n by default is 7days).

This PR solves the open issue: https://github.com/jaegertracing/helm-charts/issues/57

### CODE REVIEW

**Comment by mehta-ankit** [comment]
testing it locally right now to see this works as expected.

**Comment by naseemkullah** [comment]
👏 really nice @mehta-ankit 

Could we call the component `esIndexCleaner` to match the operator https://www.jaegertracing.io/docs/1.17/operator/#elasticsearch-index-cleaner-job
as well as the image name ?

File: charts/jaeger/README.md [hunk 1]
```diff
@@ -326,6 +326,15 @@ The following table lists the configurable parameters of the Jaeger chart and th
 | `spark.tag` | Tag of the dependencies job image | `latest` |
 | `spark.extraConfigmapMounts` | Additional spark configMap mounts | `[]` |
 | `spark.extraSecretMounts` | Additional spark secret mounts | `[]` |
+| `esCleanup.enabled` | Enables the ElasticSearch indices cleanup job| `false` |
+| `esCleanup.image` | Image for the ElasticSearch indices cleanup job| `jaegertracing/jaeger-es-index-cleaner` |
+| `esCleanup.pullPolicy` | Image pull policy of the ES cleanup image | `Always` |
+| `esCleanup.schedule` | Schedule of the cron job | `"55 23 * * *"` |
```
Comment (matthiaslee) [inline charts/jaeger/README.md]: I think `numberOfDays` is missing here

Comment (mehta-ankit) [reply charts/jaeger/README.md]: Good catch, you are right. Will add it.

**Comment by mehta-ankit** [comment]
> 👏 really nice @mehta-ankit
> 
> Could we call the component `esIndexCleaner` to match the operator https://www.jaegertracing.io/docs/1.17/operator/#elasticsearch-index-cleaner-job
> as well as the image name ?

Yes, can change the component name @naseemkullah

**Comment by mehta-ankit** [comment]
@naseemkullah I haven't run it locally yet, but just curious if you feel this error
`{"level":"info","ts":1582835705.8783262,"caller":"grpc/clientconn.go:1283","msg":"grpc: addrConn.createTransport failed to connect to {jaeger-jumo66q3ga-collector:14250 0  <nil>}. Err :connection error: desc = \"transport: Error while dialing dial tcp 10.107.199.158:14250: connect: connection refused\". Reconnecting...","system":"grpc","grpc_log":true}`
On Lint and Test Charts job an intermittent thing 🤔  ? Since by default it should not even be. deploying the CronJob so not sure how it could have affected it.

**Comment by naseemkullah** [comment]
HI @mehta-ankit there are intermittent issues with the CI 😞 I'll try re running the check

**Comment by naseemkullah** [comment]
They seem to all be failing FWIW https://github.com/jaegertracing/helm-charts/pulls

**Comment by mehta-ankit** [comment]
> They seem to all be failing FWIW https://github.com/jaegertracing/helm-charts/pulls

I see. good to know that.

**Comment by mehta-ankit** [comment]
@naseemkullah I added another commit, to add securityContext for specifying `nonRootUser` to the cronjob container. It is required by the container and it would throw an error without it. Hence adding the default value as well. Can you review that commit please.

File: charts/jaeger/templates/cleanup-es-indexes-cronjob.yaml [hunk 2]
```diff
@@ -28,8 +28,12 @@ spec:
             {{- end }}
         spec:
           serviceAccountName: {{ template "jaeger.esIndexCleaner.serviceAccountName" . }}
+          securityContext:
+{{ toYaml .Values.esIndexCleaner.podSecurityContext | indent 12 }}
           containers:
           - name: {{ include "jaeger.fullname" . }}-es-cleanup
```
Comment (naseemkullah) [inline charts/jaeger/templates/cleanup-es-indexes-cronjob.yaml]:
```suggestion
          - name: {{ include "jaeger.fullname" . }}-es-index-cleaner
```

nit: just to match the component name

File: charts/jaeger/templates/_helpers.tpl [hunk 3]
```diff
@@ -73,6 +73,17 @@ Create the name of the spark service account to use
 {{- end -}}
 {{- end -}}
 
+{{/*
+Create the name of the esIndexCleaner service account to use
+*/}}
+{{- define "jaeger.esIndexCleaner.serviceAccountName" -}}
+{{- if .Values.esIndexCleaner.serviceAccount.create -}}
+    {{ default (printf "%s-esIndexCleaner" (include "jaeger.fullname" .)) .Values.esIndexCleaner.serviceAccount.name }}
```
Comment (naseemkullah) [inline charts/jaeger/templates/_helpers.tpl]:
```suggestion
    {{ default (printf "%s-es-index-cleaner" (include "jaeger.fullname" .)) .Values.esIndexCleaner.serviceAccount.name }}
```

Comment (naseemkullah) [reply charts/jaeger/templates/_helpers.tpl]: k8s resources should be kebab-case (or whatever we call lower case + hyphens)

File: charts/jaeger/templates/cleanup-es-indexes-cronjob.yaml (hunk 2 above)
Comment (naseemkullah) [reply charts/jaeger/templates/cleanup-es-indexes-cronjob.yaml]: Could you please adjust for all instanced of `{{ include "jaeger.fullname" . }}-es-cleanup` ? 🙏

File: charts/jaeger/templates/cleanup-es-indexes-cronjob.yaml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  successfulJobsHistoryLimit: {{ .Values.esIndexCleaner.successfulJobsHistoryLimit }}
+  failedJobsHistoryLimit: {{ .Values.esIndexCleaner.failedJobsHistoryLimit }}
+  suspend: false
+  jobTemplate:
+    spec:
+      template:
+        metadata:
+          labels:
+            {{- include "jaeger.selectorLabels" . | nindent 12 }}
+            app.kubernetes.io/component: spark
```
Comment (naseemkullah) [inline charts/jaeger/templates/cleanup-es-indexes-cronjob.yaml]:
```suggestion
            app.kubernetes.io/component: es-index-cleaner
```

**Comment by naseemkullah** [comment]
Component name should be in kebab-case as well, camelCase just for helm values

File: charts/jaeger/templates/cleanup-es-indexes-cronjob.yaml [hunk 5]
```diff
@@ -0,0 +1,70 @@
+{{- if .Values.esIndexCleaner.enabled -}}
+apiVersion: batch/v1beta1
+kind: CronJob
+metadata:
+  name: {{ include "jaeger.fullname" . }}-esIndexCleaner
+  labels:
+    {{- include "jaeger.labels" . | nindent 4 }}
+    app.kubernetes.io/component: esIndexCleaner
```
Comment (naseemkullah) [inline charts/jaeger/templates/cleanup-es-indexes-cronjob.yaml]:
```suggestion
    app.kubernetes.io/component: es-index-cleaner
```

Comment (mehta-ankit) [reply charts/jaeger/templates/cleanup-es-indexes-cronjob.yaml]: Done!

File: charts/jaeger/templates/cleanup-es-indexes-cronjob.yaml (hunk 4 above)
Comment (mehta-ankit) [reply charts/jaeger/templates/cleanup-es-indexes-cronjob.yaml]: Done

**Comment by mehta-ankit** [comment]
@naseemkullah Made the changes suggested by you. Will update with my test results tomorrow, once i confirm indexes get deleted.

**Comment by naseemkullah** [comment]
Thanks @mehta-ankit !

**Comment by mehta-ankit** [comment]
@naseemkullah Test was successful (the cronjob ran and and indices were deleted).
The logs for the pod had that info as well:

image

**Comment by naseemkullah** [comment]
Seems like CI is experiencing: actions/checkout#23
Could you please try:

```
git commit --allow-empty -m "rerun build"
```
?
Might help. 🤷‍♂

**Comment by mehta-ankit** [comment]
> Seems like CI is experiencing: [actions/checkout#23](https://github.com/actions/checkout/issues/23)
> Could you please try:
> 
> ```
> git commit --allow-empty -m "rerun build"
> ```
> 
> ?
> Might help. 🤷‍♂

Trying it. Also had to resolve merge conflicts one more time (😆 ) lots of PR's merging i believe (version went up from 0.19 to 0.22) 👍 
Let's see if CI passes this time.

**Comment by mehta-ankit** [comment]
@naseemkullah Ready for merge! 🎉 CI is Green!

File: charts/jaeger/templates/_helpers.tpl (hunk 3 above)
Comment (naseemkullah) [inline charts/jaeger/templates/_helpers.tpl]:
```suggestion
    {{ default (printf "%s-es-index-cleaner" (include "jaeger.fullname" .)) .Values.esIndexCleaner.serviceAccount.name }}
```

Service account name should be in kebab-case.

File: charts/jaeger/templates/cleanup-es-indexes-cronjob.yaml [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            app.kubernetes.io/component: es-index-cleaner
+            {{- if .Values.esIndexCleaner.podLabels }}
+{{ toYaml .Values.esIndexCleaner.podLabels | indent 12 }}
+            {{- end }}
+        spec:
+          serviceAccountName: {{ template "jaeger.esIndexCleaner.serviceAccountName" . }}
+          securityContext:
+{{ toYaml .Values.esIndexCleaner.podSecurityContext | indent 12 }}
+          containers:
+          - name: {{ include "jaeger.fullname" . }}-esIndexCleaner
```
Comment (naseemkullah) [inline charts/jaeger/templates/cleanup-es-indexes-cronjob.yaml]:
```suggestion
          - name: {{ include "jaeger.fullname" . }}-es-index-cleaner
```

container name should be in kebab-cases

Comment (mehta-ankit) [reply charts/jaeger/templates/cleanup-es-indexes-cronjob.yaml]: Man, i thought i changed it everywhere 😢

Comment (mehta-ankit) [reply charts/jaeger/templates/cleanup-es-indexes-cronjob.yaml]: Made the change 🤞

**Comment by naseemkullah** [comment]
One last nit and I think we are good to go: Could you please rename the file `cleanup-es-indexes-cronjob.yaml ` to `es-index-cleaner-cronjob.yaml` ?

**Comment by mehta-ankit** [comment]
> One last nit and I think we are good to go: Could you please rename the file `cleanup-es-indexes-cronjob.yaml ` to `es-index-cleaner-cronjob.yaml` as to match the component name?

Good suggestion! Makes sense to me. 👍  (will remember it next time)

**Comment by mehta-ankit** [comment]
@naseemkullah Ready for merge!

**Comment by naseemkullah** [comment]
Thanks @mehta-ankit !

**Comment by PardisTV** [comment]
hi @mehta-ankit , can we do the same thing when we are using badger instead of elasticsearch?
I mean having a cronjob to purge Jaeger's data after 30 days?

**Comment by mehta-ankit** [comment]
> hi @mehta-ankit , can we do the same thing when we are using badger instead of elasticsearch? I mean having a cronjob to purge Jaeger's data after 30 days?

@PardisTV I am sure you can do that, although since I do not use badger I do not know how. 
Also can you open an issue and we can discuss it there, rather than discussing this on a closed PR. Thanks

**Pull request merged**